File: winekloo/frontend/lib/models/restaurateur_model.dart
import 'package:flutter/material.dart';

class Restaurateur {
  final int? restaurateurID;
  final String? diningType;
  final String? name;
  final String? email;
  final String? phoneNumber;
  final String? password;
  final String? location;
  final String? photo;
  final Map<String, Map<String, String>>? workingHours;
  final List<String>? workingDays;
  final String? description;
  final double? ratingValueAverage;
  final String? menuFilePDF;
  final int? restaurateurCategories;
  final int? restaurateurPreferences;
  final int? restaurateurSpecialFeatures;
  final int? restaurateurPricing;

  Restaurateur({
    this.restaurateurID,
    this.diningType,
    this.name,
    this.email,
    this.phoneNumber,
    this.password,
    this.location,
    this.photo,
    this.workingHours,
    this.workingDays,
    this.description,
    this.ratingValueAverage,
    this.menuFilePDF,
    this.restaurateurCategories,
    this.restaurateurPreferences,
    this.restaurateurSpecialFeatures,
    this.restaurateurPricing,
  });

  factory Restaurateur.fromJson(Map<String, dynamic> json) {
    return Restaurateur(
      restaurateurID: json['RestaurateurID'],
      diningType: json['DiningType'],
      name: json['Name'],
      email: json['Email'],
      phoneNumber: json['PhoneNumber'],
      password: json['Password'],
      location: json['Location'],
      photo: json['Photo'],
      workingHours: (json['WorkingHours'] as Map<String, dynamic>?)?.map(
      (key, value) => MapEntry(
        key,
        (value as Map<String, dynamic>).map(
          (innerKey, innerValue) => MapEntry(innerKey, innerValue as String),
        ),
      ),
    ),
      workingDays: (json['WorkingDays'] as List<dynamic>?)?.cast<String>(),
      description: json['Description'],
      ratingValueAverage: json['RatingValueAverage']?.toDouble(),
      menuFilePDF: json['MenuFilePDF'],
      restaurateurCategories: json['Categories'],
      restaurateurPreferences: json['DietaryPreferences'],
      restaurateurSpecialFeatures: json['SpecialFeatures'],
      restaurateurPricing: json['Pricing'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'RestaurateurID': restaurateurID,
      'DiningType': diningType,
      'Name': name,
      'Email': email,
      'PhoneNumber': phoneNumber,
      'Password': password,
      'Location': location,
      'Photo': photo,
      'WorkingHours': workingHours,
      'WorkingDays': workingDays,
      'Description': description,
      'RatingValueAverage': ratingValueAverage,
      'MenuFilePDF': menuFilePDF,
      'Categories': restaurateurCategories,
      'DietaryPreferences': restaurateurPreferences,
      'SpecialFeatures': restaurateurSpecialFeatures,
      'Pricing': restaurateurPricing,
    };
  }

  static TimeOfDay _timeOfDayFromString(String timeString) {
    final timeParts = timeString.split(':');
    return TimeOfDay(
      hour: int.parse(timeParts[0]),
      minute: int.parse(timeParts[1]),
    );
  }

  static String _timeOfDayToString(TimeOfDay time) {
    return '${time.hour.toString().padLeft(2, '0')}:${time.minute.toString().padLeft(2, '0')}';
  }
}

File: winekloo/backend/app/models/restaurateur_model.py
from app.db import db

class Restaurateur(db.Model):
    __tablename__ = 'restaurateur'

    restaurateur_id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    dining_type = db.Column(db.String(255), nullable=True)
    name = db.Column(db.String(255), nullable=True)
    email = db.Column(db.String(255), unique=True, nullable=False)
    phone_number = db.Column(db.String(20), unique=True, nullable=False)
    password = db.Column(db.String(255), nullable=False)
    location = db.Column(db.String(255), nullable=True)
    photo = db.Column(db.String(255), nullable=True)
    working_hours = db.Column(db.JSON, nullable=True)
    working_days = db.Column(db.JSON, nullable=True)
    description = db.Column(db.String(1000), nullable=True)
    rating_value_average = db.Column(db.Float, nullable=True)
    menu_file_pdf = db.Column(db.String(255), nullable=True)
    restaurateur_categories = db.Column(db.Integer, nullable=True)
    restaurateur_preferences = db.Column(db.Integer, nullable=True)
    restaurateur_special_features = db.Column(db.Integer, nullable=True)
    restaurateur_pricing = db.Column(db.Integer, nullable=True)

    def __init__(self, email, phone_number, password, dining_type=None, name=None, location=None, photo=None,
                 working_hours=None, working_days=None, description=None, rating_value_average=None,
                 menu_file_pdf=None, restaurateur_categories=None, restaurateur_preferences=None,
                 restaurateur_special_features=None, restaurateur_pricing=None):
        self.email = email
        self.phone_number = phone_number
        self.password = password
        self.dining_type = dining_type
        self.name = name
        self.location = location
        self.photo = photo
        self.working_hours = working_hours
        self.working_days = working_days
        self.description = description
        self.rating_value_average = rating_value_average
        self.menu_file_pdf = menu_file_pdf
        self.restaurateur_categories = restaurateur_categories
        self.restaurateur_preferences = restaurateur_preferences
        self.restaurateur_special_features = restaurateur_special_features
        self.restaurateur_pricing = restaurateur_pricing

    @staticmethod
    def from_json(json_data):
        return Restaurateur(
            restaurateur_id=json_data.get('RestaurateurID'),
            email=json_data.get('Email'),
            phone_number=json_data.get('PhoneNumber'),
            password=json_data.get('Password'),
            dining_type=json_data.get('DiningType'),
            name=json_data.get('Name'),
            location=json_data.get('Location'),
            photo=json_data.get('Photo'),
            working_hours=json_data.get('WorkingHours'),
            working_days=json_data.get('WorkingDays'),
            description=json_data.get('Description'),
            rating_value_average=json_data.get('RatingValueAverage'),
            menu_file_pdf=json_data.get('MenuFilePDF'),
            restaurateur_categories=json_data.get('Categories'),
            restaurateur_preferences=json_data.get('DietaryPreferences'),
            restaurateur_special_features=json_data.get('SpecialFeatures'),
            restaurateur_pricing=json_data.get('Pricing')
        )

    def to_json(self):
        return {
            'RestaurateurID': self.restaurateur_id,
            'Email': self.email,
            'PhoneNumber': self.phone_number,
            'Password': self.password,
            'DiningType': self.dining_type,
            'Name': self.name,
            'Location': self.location,
            'Photo': self.photo,
            'WorkingHours': self.working_hours,
            'WorkingDays': self.working_days,
            'Description': self.description,
            'RatingValueAverage': self.rating_value_average,
            'MenuFilePDF': self.menu_file_pdf,
            'Categories': self.restaurateur_categories,
            'DietaryPreferences': self.restaurateur_preferences,
            'SpecialFeatures': self.restaurateur_special_features,
            'Pricing': self.restaurateur_pricing
        }
